contracts/libraries/SignatureVerifier.sol

```diff
@@ -0,0 +1,214 @@
+    ) public view override returns (address signer) {
+        bytes32 loanHash = keccak256(
+            abi.encode(
+                _BUNDLE_TYPEHASH,
+                loanTerms.durationSecs,
+                loanTerms.principal,
+                loanTerms.interest,
+                loanTerms.bundleId,
+                loanTerms.payableCurrency
+            )
+        );
+
+        bytes32 typedLoanHash = _hashTypedDataV4(loanHash);
+        signer = ECDSA.recover(typedLoanHash, v, r, s);
+    }
+
+    /**
+     * @notice Determine the external signer for a signature specifying specific items.
+     * @dev    Bundle ID should _not_ be included in this signature, because the loan
+     *         can be initiated with any arbitrary bundle - as long as the bundle contains the items.
+     *
+     * @param loanTerms                     The terms of the loan.
+     * @param v                             Part of the signature.
+     * @param r                             Part of the signature.
+     * @param s                             Part of the signature.
+     * @param items                         The required items in the specified bundle.
+     *
+     * @return signer                       The address of the recovered signer.
+     */
+    function recoverItemsSignature(
+        LoanLibrary.LoanTerms calldata loanTerms,
+        uint8 v,
+        bytes32 r,
+        bytes32 s,
+        bytes calldata items
+    ) public view override returns (address signer) {
+        bytes32 loanHash = keccak256(
+            abi.encode(
+                _ITEMS_TYPEHASH,
+                loanTerms.durationSecs,
+                loanTerms.principal,
+                loanTerms.interest,
+                items,
+                loanTerms.payableCurrency
+            )
+        );
+
+        bytes32 typedLoanHash = _hashTypedDataV4(loanHash);
+        signer = ECDSA.recover(typedLoanHash, v, r, s);
+    }
+
+    // ==================================== COLLATERAL VERIFICATION =====================================
+
+    /**
+     * @notice Verify that the items specified by the packed SignatureItem array are held by the vault.
+     * @dev    Reverts on a malformed SignatureItem, returns false on missing contents.
+     *
+     * @param itemsPacked                   The SignatureItem[] array of items, packed in bytes.
+     * @param vault                         The vault that should own the specified items.
+     *
+     * @return verified                     Whether the bundle contains the specified items.
+     */
+    // solhint-disable-next-line code-complexity
+    function verifyItems(
+        bytes calldata itemsPacked,
+        address vault
+    ) public view override returns (bool) {
+        // Unpack items
+        (LoanLibrary.SignatureItem[] memory items) = abi.decode(itemsPacked, (LoanLibrary.SignatureItem[]));
+
+        for (uint256 i = 0; i < items.length; i++) {
+            LoanLibrary.SignatureItem memory item = items[i];
```

Comment:
`verifyItems` is non-deterministic in its gas consumption due to unknown size of `items` at runtime. Would this memory allocation be more gas efficient converting to inline assembly ?